Streamline EKS BPF enablement (#4258)
projectcalico/calico: maintenance/ebpf/ebpf-and-eks.md

### How to

> #### Create an EKS cluster with a recent enough kernel

Reviewer: Not sure what "recent enough" means...

Why not just "Create an EKS cluster"? The requirement is a kernel version, which we explain so not sure we need to qualify this in the title.

Author: How about:
```suggestion
#### Create an eBPF compatible EKS cluster 
```

<label:Bottlerocket,active:true>
<%

The easiest way to start an EKS cluster that meets eBPF mode's requirements is to use Amazon's 
[Bottlerocket](https://aws.amazon.com/bottlerocket/) OS, instead of the default.  Bottlerocket is a 

> container-optimised OS with an emphasis on security; it has a recent enough kernel to use eBPF mode.

Reviewer: it has a version of the kernel works for eBPF mode.

Author: How about: 
```suggestion
container-optimised OS with an emphasis on security; it has a version of the kernel which is compatible with eBPF mode.
```

    ```bash
    kubectl apply -f {{ "/manifests/calico-vxlan.yaml" | absolute_url }}
    ```

* Create a nodegroup, using the AMI ID you noted above.
  * `--node-ami` should be set to the AMI ID of the image built above.
  * `--node-ami-family` should be set to `Ubuntu1804` (despite the upgrade).

  For example:
  ```
  eksctl create nodegroup \
    --cluster my-calico-cluster \
    --node-type t3.medium \
    --node-ami auto \
    --max-pods-per-node 100 \
    --node-ami-family Ubuntu1804 \
    --node-ami <AMI ID>
  ```

%>
{% endtabs %}

> #### Configure {{site.prodname}} to connect directly to the API server

Reviewer: Configure {{site.prodname}} to connect to the API server

Author: "Directly" is important here - before kube-proxy is disabled, calico-node can already connect to the API server via the service IP set up by kube-proxy.  

Once we delete kube-proxy, calico-node (and typha, calico-kube-controllers) will need to connect to the API server directly - i.e. without any assistance from kube-proxy.

Reviewer: Got it. Ignore

  and `443` (the default for HTTPS) for `KUBERNETES_SERVICE_PORT` when creating the config map.

  Create the following config map in the `kube-system` namespace using the host and port determined above:

  ```
  kubectl apply -f - <<EOF
  kind: ConfigMap
  apiVersion: v1
  metadata:
    name: kubernetes-services-endpoint
    namespace: kube-system
  data:
    KUBERNETES_SERVICE_HOST: "<API server host>"
    KUBERNETES_SERVICE_PORT: "443"
  EOF
  ```

* Wait 60s for kubelet to pick up the `ConfigMap` (see Kubernetes [issue #30189](https://github.com/kubernetes/kubernetes/issues/30189){:target="_blank"}); then, restart the {{site.prodname}} pods to pick up the change:

  ```
  kubectl delete pod -n kube-system -l k8s-app=calico-node
  ```

  And, if using Typha and/or calico-kube-controllers (if you're not sure if you're running these, run the commands
  anyway, they will fail with "No resources found" if the pods aren't present):
  ```
  kubectl delete pod -n kube-system -l k8s-app=calico-typha
  kubectl delete pod -n kube-system -l k8s-app=calico-kube-controllers
  ```

> * Confirm that pods restart and then reach the `Running` state with the following command:

Reviewer: Confirm that pods restart and reach the `Running` state with the following command:

  ```
  watch "kubectl get pods -n kube-system | grep calico"
  ```

>   You can verify that the change was picked up by checking the logs of one of the  {{ site.nodecontainer }} pods.

Reviewer: extra space before {{ site.nodecontainer }}

  ```
  kubectl get po -n kube-system -l k8s-app=calico-node
  ```

  Should show one or more pods:

  ```
  NAME                                       READY   STATUS    RESTARTS   AGE
  {{site.noderunning}}-d6znw                          1/1     Running   0          48m
  ...
  ```

  Then, to search the logs, choose a pod and run:

  ```
  kubectl logs -n kube-system <pod name> | grep KUBERNETES_SERVICE_HOST
  ```

  You should see the following log, with the correct `KUBERNETES_SERVICE_...` values.

  ```
  2020-08-26 12:26:29.025 [INFO][7] daemon.go 182: Kubernetes server override env vars. KUBERNETES_SERVICE_HOST="172.16.101.157" KUBERNETES_SERVICE_PORT="6443"
  ```

#### Disable kube-proxy

> 

Reviewer: In eBPF mode, {{site.prodname}} ...

In eBPF mode {{site.prodname}} replaces `kube-proxy` so it wastes resources to run both.  To disable `kube-proxy` reversibly, we recommend adding a node selector to `kube-proxy`'s `DaemonSet` that matches no nodes, for example:

```
kubectl patch ds -n kube-system kube-proxy -p '{"spec":{"template":{"spec":{"nodeSelector":{"non-calico": "true"}}}}}'
```

Then, should you want to start `kube-proxy` again, you can simply remove the node selector.

#### Enable eBPF mode

> To enable eBPF mode, change Felix configuration parameter  `BPFEnabled` to `true`.  This can be done with `calicoctl`, as follows:

Reviewer: To enable eBPF mode, change the Felix configuration parameter `BPFEnabled` to `true`.

```
calicoctl patch felixconfiguration default --patch='{"spec": {"bpfEnabled": true}}'
```

Enabling eBPF node can disrupt existing workload connections.  After enabling eBPF mode you may need to restart
workload pods in order for them to restart connections.  In particular, it's a good idea to restart `kube-dns`
since its connection to the API server can be disrupted:

```
kubectl delete pod -n kube-system -l k8s-app=kube-dns
```

#### Disable eBPF mode

To revert to standard Linux networking:

1. Disable Calico eBPF mode:

   ```
   calicoctl patch felixconfiguration default --patch='{"spec": {"bpfEnabled": false}}'
   ```

1. If you disabled `kube-proxy`, re-enable it (for example, by removing the node selector added above).
   ```
   kubectl patch ds -n kube-system kube-proxy --type merge -p '{"spec":{"template":{"spec":{"nodeSelector":{"non-calico": null}}}}}'
   ```

> 1. Monitor existing workloads to make sure they re-establish any connections disrupted by the switch.

Reviewer: Monitor existing workloads to make sure they reestablish connections.
